docs/scripts/patch_devsite_toc.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+python-bigtable is made up of multiple clients, which we want to categorize separately in
+the devsite table of contents.
+
+This script allows us to add separate sections to the autogenerated toc.yml file to
+accomplish this.
+"""
+
+
+import yaml
+import os
+import shutil
+
+# set working directory to /docs
+os.chdir(f"{os.path.dirname(os.path.abspath(__file__))}/{os.pardir}")
+
+
+def add_sections(toc_file_path, section_list, output_file_path=None):
+    """
+    Add new sections to the autogenerated docfx table of contents file
+
+    Takes in a list of TocSection objects, which should point to a directory of rst files
+    within the main /docs directory, which represents a self-contained section of content
+
+    :param toc_file_path: path to the autogenerated toc file
+    :param section_list: list of TocSection objects to add
+    :param output_file_path: path to save the updated toc file. If None, save to the input file
+    """
+    current_toc = yaml.safe_load(open(toc_file_path, "r"))
+    for section in section_list:
+        print(f"Adding section {section.title}...")
+        current_toc[0]["items"].insert(-1, section.to_dict())
+        section.copy_markdown()
+    # save file
+    if output_file_path is None:
+        output_file_path = toc_file_path
+    with open(output_file_path, "w") as f:
+        yaml.dump(current_toc, f)
+
+
+class TocSection:
+    def __init__(self, dir_name, index_file_name):
+        """
+        :param dir_name: name of the directory containing the rst files
+        :param index_file_name: name of an index file within dir_name. This file
+            will not be included in the table of contents, but provides an ordered
+            list of the other files which should be included
+        """
+        self.dir_name = dir_name
+        self.index_file_name = index_file_name
+        index_file_path = os.path.join(dir_name, index_file_name)
+        # find set of files referenced by the index file
+        with open(index_file_path, "r") as f:
+            self.title = f.readline().replace("\n", "")
+            in_toc = False
+            self.items = []
+            for line in f:
```

Comment:
I think this can be easier to read if flattened a bit:

```python
for line in f:
  # ignore empty lines
  if not line.rstrip('\n'):
     continue
  if line.startswith(".. toctree::"):
     in_toc = True
     continue
  # ignore directives
  if ":" in line:
     continue
  if not in_toc:
    continue
  # bail when toc indented block is done
  if not line.startswith(" ") or not line.startswith("\t"):
    # done with indented toc
    break
  # extract entries
  self.items.append(extract_toc_entry(line.strip())
```